Fix crash when deleting with multiple cursors

pascalkuthe opened the pull request:
Fixes #6021
Fixes https://github.com/helix-editor/helix/issues/5272
Fixes https://github.com/helix-editor/helix/issues/6029

The `delete_char_backward` command delete indentation when at the start of the line. The problem is that this can cause multiple overlaing deletions with multiple cursors. For example in #6021 one of the cursors is after the 3. space and one after the 1. char. Because the indent unit is 4 spaces `delete_char_backward` deletes to the start of the line for both cursors. The generated transaction is therefore invalid which causes an integer underflow when the transaction is applied.

I fixed a similar crash in #5728. Mutlicursors potentially generating overlapping changes is a common pitfall and something we should be wary of. In that PR I simply discarded changes that overlapped a previous change. That seemed the right behaviour to me for the **replacements** created by completions. For **deletions** however we can simply merge the overlapping deletions together (otherwise the behaviour here would be incorrect). To that end I added a new `delete` and `delete_by_selection` function which does not accept a tendril for replacement and allows overlapping ranges.

In the process I noticed that the `delete_by_selection_insert_mode` function was implemented a bit weirdly:
* It requires a selection rather then a closure like `Transaction::change_by_selection` which requires cloneing the selection
* Calles generating transaction causes some weirdlooking code
* It doesn't automatically invoke `lsp::signature_help_impl(cx, SignatureHelpInvoked::Automatic);` eventough all callsites do (wihout doing that its barely wort moving the logic to a shared function)

Therefore I cleaned up that function a little in a second commit as I was touching that code anyways (and there were no open PRs that touch that code).
If we don't want these cosmetic changes I can easily drop that second commit.

gabydd commented:
thanks for fixing this Pascal, it also fixes https://github.com/helix-editor/helix/issues/5272

pascalkuthe commented:
> thanks for fixing this Pascal, it also fixes #5272

Thanks for confirming that :+1: That issue was on my todo too so I get two for one here :smile:

pascalkuthe commented:
I added a fix for #6029 here as the refactor in [3edcc5d](https://github.com/helix-editor/helix/pull/6024/commits/3edcc5dc3df58aeca1d760813a5052c7870e60ac) made that easier to implement.
I added a detailed explanation in the commit message of [f0f2dae](https://github.com/helix-editor/helix/pull/6024/commits/f0f2daef76bcf89e34acd035948817601bc739e5).

The TLDR is:

Deleting forward in append mode used to delete one char forward and then delete backwards, because the selection was shrunk. I have changed the command to always delete forward. The following example of pressing `<del>` repeatedly in append mode explains it well I think.

What I expect to happen (and how it behaves with this PR):
```
|abcd|ef
|abce|f
|abcf|
|abc |
```

This is inline with how other editors like kakoune work.
However, helix currently moves the selection backwards leading to the
following behavior:

```
|abcd|ef
|abc|ef
|ab|ef
|e|f
```

the-mikedavis commented on helix-core/src/transaction.rs:
```diff
@@ -534,6 +535,46 @@ impl Transaction {
         Self::from(changeset)
     }
 
+    /// Generate a transaction from a set of potentially overallping deletions
```
```suggestion
    /// Generate a transaction from a set of potentially overlapping deletions
```

the-mikedavis commented on helix-core/src/transaction.rs:
```diff
@@ -534,6 +535,46 @@ impl Transaction {
 
+    /// Generate a transaction from a set of potentially overallping deletions
+    /// by merging overlapping deletions together.
+    pub fn delete<I>(doc: &Rope, changes: I) -> Self
```
very minor nit but I think this would be clearer as `deletions` rather than `changes`

pascalkuthe replied:
I have fixed that in the force push

the-mikedavis commented on helix-term/src/commands.rs:
```diff
@@ -2325,11 +2320,49 @@ fn delete_selection_impl(cx: &mut Context, op: Operation) {
+                    insert_newline |= end == text_len;
+                    // move the cursor to the right so that the selection
+                    // doesn't shrink when deleting forward (so the text appears to
+                    // move the the left)
+                    // += 1 is enoug here as the range is normalized to grapheme boundaries
+                    // later anyway
```
```suggestion
                    // move to the left)
                    // += 1 is enough here as the range is normalized to grapheme boundaries
                    // later anyway
```

the-mikedavis reviewed (commented):
Just a few minor comments, otherwise I think this looks good

pascalkuthe commented:
(force pushed to keep the history intact since I did put some effort into commit messages here :D)

dead10ck reviewed (commented):
Ok the changes themselves look good to me. I would humbly request regression tests to codify expected behavior, but not sure if you or others feel this is worth blocking on.

I spent a little time thinking over the larger API—it doesn't quite feel right to fix the function that generates invalid transactions by adding another function that doesn't and changing specific call sites to that instead. But looking over the `change` function, I can see why we can't just move the `from` marker forward in the case of overlap, since this could just end up inserting text in the wrong place.

My first instinct was to suggest that we either

1. Just keep adding and deleting text separate, or
2. Fix this issue in `change` and just make `delete` a thin wrapper around it

to fix even the possibility of this happening through the API. But 1 would make cases like `R` very awkward, and 2 I can't think of a reasonable way to handle overlapping ranges when it handles both deletions and text replacement, if it's even possible, without making it super complex.

So at the end of the day, having another function that relaxes the overlapping ranges invariant for text deletions gets the job done.

pascalkuthe commented:
@dead10ck I add some integration tests that was a great suggestions. We should really make more use of the integration test framework it works really great. I was pretty much just able to use the reproduction cases straight from the issues.

Regarding the API I actually think having a second function makes sense. For normal edits (replacements) there is no unambiguous way to merge overlapping edits. Instead called must deal with that (or we panic).  There is no correct way to handle this generally in `change`. Even separating the deleting and adding doesn't really fix the problem. The callers would map multiple positions to the same location in that case but there is still no unambiguous way to now add that new text (so again we would have just turned an explicit panic into weirdly/silently incorrect changes).

I spent a lot of time thinking about a better API here when I created this PR but I came to the conclusion that there really is none. Overlapping edits are fundamentally incorrect and must be prevented in the caller. The only exceptions are:

* deletions (where you can just merge the ranges as done here)
* "foreign" edits (like completions) over which we have no control. There we simply use the first range and discard all subsequent overlapping ranges (a decent fallback strategy but this doesn't work for deletions and some other cases)

When I created this PR I actually manually went trough all callsites to the `change`/`change_by_selection` and manually checked that this can't be a problem there. As far as I found all callsites were correct (in most cases edit ranges are just completions in which case overlapping edits are impossible and in the cases where they were possible the functions had explicit checks to deal with that). Only deletions were buggy.

pascalkuthe commented on helix-core/src/transaction.rs:
```diff
@@ -534,6 +535,41 @@ impl Transaction {
+                from = last
+            }
+            debug_assert!(
+                from <= to,
```
this is actually a copy of an existing assertion from the `change` function (the function in general is almost fully copied). If we change this here we should change it there too. I guess the point of this is to basically get a better panic message in debug builds (and to just let it underflow in release builds)

dead10ck commented on helix-term/tests/test/commands.rs:
```diff
@@ -385,3 +385,70 @@ async fn test_character_info() -> anyhow::Result<()> {
+async fn test_delete_char_backward() -> anyhow::Result<()> {
+    // don't panic when deleting overlapping ranges
+    test((
+        platform_line("#(x|)# #[x|]#").as_str(),
```
The `as_str` actually shouldn't be necessary any more after 5b07ce7

pascalkuthe replied:
I took the chance to fix this for all tests in that file and therefore made this a separate commit [d605a34](https://github.com/helix-editor/helix/pull/6024/commits/d605a341073325b7f406db524a246e756498c51a)

dead10ck commented on helix-term/tests/test/commands.rs:
```diff
@@ -385,3 +385,70 @@ async fn test_character_info() -> anyhow::Result<()> {
+    test((
+        platform_line("#( |)##( |)#a#( |)#axx#[x|]#a").as_str(),
+        "li<backspace><esc>",
+        platform_line("#(a|)##(|a)#xx#[|a]#").as_str(),
```
Why is the first range forward while the rest are backward? How does this test pass? Is there a bug here where the order of the ranges isn't being preserved?

pascalkuthe replied:
This is just how range mapping in helix works. We basically delete all chars inside the selection which yields an empty selection. Helix normalizes this to a single char selection but during that normalization the selection direction is lost. For single char selections the direction doesn't really matter tough so it's not usually noticabke in the ui.

This has nothing to do withthese specific commands tough as they dont actually touch the selection at all, it's just how the automatic selection mapping works.

dead10ck replied:
Interesting, ok, thanks for the explanation

dead10ck commented on helix-term/src/commands.rs:
```diff
@@ -2325,11 +2320,49 @@ fn delete_selection_impl(cx: &mut Context, op: Operation) {
+    if direction == Direction::Forward {
+        doc.set_selection(
+            view.id,
+            Selection::new(selection, doc.selection(view.id).primary_index()),
```
If ranges can potentially get merged, can we always use the original selection's primary index safely?

pascalkuthe replied:
Its save here because we just delete the selections exactly (start/end is always just the start/end of the selection range). That means that this particular function can actually never have overlapping deletion ranges (since selections are always mon-overlapping).

pascalkuthe replied:
Ah my bad aplogoues I looked at the wrong code somehow. This is a trick. Here the selection actually is set manually. This works because we call `set_selection` *before* the transaction is applied and if you look at the code we unconditionally call push so the number of ranges doesn't change and the primary index remains valid. 

We only update the selection this way (with head += 1) so that the semantics for `delete_X_forward` work as intended. Just the normally mapping the selection trough the transaction. When you call set_selection I'd normalized si if the selections overlap (because of the head += 1) that's automatically taken care of (and the primary index is then adjusted automatically). 

Furthermore when we apply the transaction below the new selection is still mapped trough the change set so only then becomes then do "overlapping deletions" become a problem. However as long as the selection is valid here (which it is) then that mapping will work too.

dead10ck replied:
Ahhh, I get it now, so all the overlapping selections are still actually there at the end of the function, and they get removed during the transaction application. Thanks for bearing with me.

dead10ck commented on helix-term/tests/test/commands.rs:
```diff
@@ -385,3 +385,70 @@ async fn test_character_info() -> anyhow::Result<()> {
+async fn test_delete_word_forward() -> anyhow::Result<()> {
+    // don't panic when deleting overlapping ranges
+    test((
+        platform_line("fo#[o|]#b#(|ar)#").as_str(),
```
We have the first range forward and the second backward here

pascalkuthe replied:
`i` makes both ranges forward. I can change this to make this less awkward looking but it doesn't actually affect the behaviour of the test (since all text get deleted helix always places a forward cursor on the newline regardless if what the selection was previously l

dead10ck replied:
Yeah you're right, this isn't a big deal

dead10ck commented on helix-term/tests/test/commands.rs:
```diff
@@ -385,3 +385,70 @@ async fn test_character_info() -> anyhow::Result<()> {
+        platform_line(indoc! {"\
+                #[abc|]#ef
+                #(abc|)#f
+                #(abc|)#
```
Why does `f` get deleted here? I would expect del to delete the character to the right of the selection. After entering append mode, the cursor should be over `f`, so I would expect the LF to get deleted.

pascalkuthe replied:
Del deletes the char under the cursor actually (also on master). It's the same for kakune iirc.

dead10ck reviewed (changes requested):
Thanks a lot for adding tests! This will help a lot in the future for anyone who makes changes to deleting behavior.

Also, I agree with your points about the API and adding a function. It does make sense to have both.

dead10ck reviewed (approved):
Ok, this looks good to me! Thanks for adding tests and bearing with me. This will help when I get to auto pair deletions.
